src/Pim/Bundle/DataGridBundle/Datasource/ProductDatasource.php

```diff
@@ -53,7 +65,17 @@ public function getResults()
             QueryBuilderUtility::removeExtraParameters($this->qb);
         }
 
-        $rows = $this->hydrator->hydrate($this->qb, $options);
+        $productCursor = $this->pqb->execute();
+        $context = ['locale' => $options['locale_code'], 'channel' => $options['scope_code']];
+        $rows = ['totalRecords' => $productCursor->count()];
+
+        foreach ($productCursor as $product) {
+            $normalizedProduct = array_merge(
+                $this->normalizer->normalize($product, 'internal_api', $context),
```

Comment:
And the Magic happens
That's the first real step to remove all the complexity of the grid :+1: